src/utils/cosmosdb_model_versioned.ts

```diff
@@ -0,0 +1,190 @@
+import {
+  BaseModel,
+  CosmosdbModel,
+  CosmosDecodingError,
+  CosmosErrorResponse,
+  CosmosErrors
+} from "./cosmosdb_model";
+
+import * as t from "io-ts";
+
+import { right } from "fp-ts/lib/Either";
+import { fromNullable, none, Option, some } from "fp-ts/lib/Option";
+import { fromEither, TaskEither, tryCatch } from "fp-ts/lib/TaskEither";
+
+import { PromiseType } from "italia-ts-commons/lib/types";
+
+import { NonNegativeInteger } from "italia-ts-commons/lib/numbers";
+
+import {
+  Container,
+  ErrorResponse,
+  ItemDefinition,
+  RequestOptions,
+  SqlQuerySpec
+} from "@azure/cosmos";
+import { wrapWithKind } from "./types";
+
+export const ModelId = t.string; // FIXME: make it branded
+
+export type ModelId = t.TypeOf<typeof ModelId>;
+
+/**
+ * A VersionedModel should track the version of the model
+ */
+export const VersionedModel = t.interface({
+  version: NonNegativeInteger
+});
+
+export type VersionedModel = t.TypeOf<typeof VersionedModel>;
+
+/**
+ * Returns a string with a composite id that has the format:
+ * MODEL_ID-VERSION
+ *
+ * MODEL_ID is the base model ID
+ * VERSION is the zero-padded version of the model
+ *
+ * @param modelId The base model ID
+ * @param version The version of the model
+ */
+export function generateVersionedModelId(
+  modelId: ModelId,
+  version: NonNegativeInteger
+): string {
+  const paddingLength = 16; // length of Number.MAX_SAFE_INTEGER == 9007199254740991
+  const paddedVersion = ("0".repeat(paddingLength) + String(version)).slice(
+    -paddingLength
+  );
+  return `${modelId}-${paddedVersion}`;
+}
+
+const incVersion = (version: NonNegativeInteger) =>
+  (Number(version) + 1) as NonNegativeInteger;
+
+/**
+ * Assumption: the model ID is also the partition key
+ */
+export abstract class CosmosdbModelVersioned<
+  T,
+  TN extends Readonly<T & Partial<VersionedModel>>,
+  TR extends Readonly<T & VersionedModel & BaseModel>
+> extends CosmosdbModel<T, TN & BaseModel, TR> {
+  constructor(
+    container: Container,
+    protected readonly newVersionedItemT: t.Type<TN, ItemDefinition, unknown>,
+    protected readonly retrievedItemT: t.Type<TR, unknown, unknown>,
+    protected readonly modelIdKey: keyof T
+  ) {
+    super(
+      container,
+      t.intersection([newVersionedItemT, BaseModel]),
+      retrievedItemT
+    );
+  }
+
+  public create = (
+    o: TN,
+    options?: RequestOptions
+  ): TaskEither<CosmosErrors, TR> => {
+    // if we get an explicit version number from the new document we use that,
+    // or else we set version to 0
+    const currentVersion: NonNegativeInteger | undefined = o.version;
+    const version =
+      currentVersion === undefined ? (0 as NonNegativeInteger) : currentVersion;
```

Comment:
Ensure is NonNegativeInteger

```suggestion
    const version = NonNegativeInteger.decode(o.version).getOrElse(
      0 as NonNegativeInteger
    );
```